app/controllers/pipeline_viz_controller.rb

```diff
@@ -13,21 +13,47 @@ def show
     pipeline_run = sample && sample.first_pipeline_run
 
     if feature_allowed && pipeline_run
-      stage_info = {}
-      pipeline_run.pipeline_run_stages.each do |stage|
+      @stages = []
+      @edges = []
+      all_stage_dag_jsons = []
+      pipeline_run.pipeline_run_stages.each_with_index do |stage, stage_index|
```

Comment:
A few high-level comments:
- *To keep this controller thin*, we should move this logic to a `PipelineRunGraphService` (you can change the name) that takes care of all the data structure rearrangement.
- Add comments at the start to describe the structure you are going for. This will allow us to discuss the structure before discussing the actual code.
- Should we return a single structure, e.g. graph with edges and stages, as opposed to separate params?

Reply:
I created `RetrievePipelineVizGraphDataService`, added the comment, and changed it to a single structure so the service return value is easier / cleaner